remove autoexpansion of 2-arg Mul

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
#4596 continued from #17237 

See also [this thread](https://groups.google.com/forum/#!searchin/sympy/automatic$20distribution|sort:date/sympy/MrvwPBvZIqo/ebruCXLpL0IJ) and [this one](https://groups.google.com/forum/#!searchin/sympy/2-arg%7Csort:date/sympy/5AEPk2LY9YE/t24SLaPC7mQJ).

#### Brief description of what is fixed or changed

Formerly
```python
>>> 2*(x + 1)
2*x + 2
>>> -1*(x + 1)
-x  - 1
>>> -(x + 1)
-x - 1
```
Now the input is unchanged for the first two while negation is no longer handled as multiplication by -1: 

```python
>>> 2*(x + 1)
2*(x + 1)
>>> -1*(x + 1)
-(x  + 1)
>>> -(x + 1)
-x - 1
```
#### Other comments

This allows some more meaningful/natural ways to display expressions:
```python
>>> x-y-z
x - y - z
>>> collect_const(_)
x - (y + z)
```

Some of the details raised in the discussion about this issue are handled. From the `Add.flatten` docstring:

```python
        >>> from sympy.abc import w, x, y, z
```
        A negative term will cancel a positive term
```python
        >>> a = -(x + y)
        >>> a - a
        0
```
        A negated term will have the args flattened:

```python
        >>> z - (x + y)
        -x - y + z
```
        To keep this from being flattened, explicitly
        multiply by -1:

```python
        >>> z + -1*(x + y)
        z - (x + y)
```
        Args within a negated sum will cancel with args
        that are not nested:

```python
        >>> 2*x + 2*y + -1*(w - x + 2*y + z)
        3*x - (w + z)
```
        If a term is a 2-arg Mul with coefficient other than -1,
        nothing will happen to it:

```python
        >>> x + y - 2*(x + y)
        x + y - 2*(x + y)
```

The failures of the doctests are a good place to look to generate discussion to see what other features should be supported so as few as possible major changes will need to be made. I have only worked on getting the `core` doctests to pass. Correcting everything will be like...[removing the Mind Stone from Vision's forehead](https://www.youtube.com/watch?v=lfsTCXLQU4Y). Or we may decide that this just isn't worth it.

The encouraging news so far is that `test('/core')` takes about 40% as long (**less than half as long**). 

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

<!-- END RELEASE NOTES -->

### While correcting, check for these old assumptions
- if `x` is an Add then `-x` is still an Add [handled with `neg` function]

smichr commented:
@oscarbenjamin said,

I'm running the tests locally. It looks to me like there are a lot of tests that are failing because the precise form of the output has changed:
```
>       assert dsolve(eq2) == sol2
E       assert Eq(f(x), (C1 + 5*x)/3) == Eq(f(x), C1 + 5*x/3)
```
Others are failing because the form of the input has changed and no longer matches:
```
>           raise ValueError(string + str(eq) + " does not match hint " + hint)
E           ValueError: ODE -x**2 + x*f(x) + Derivative(f(x), x) does not match hint 1st_linear
```
(that's just a bug somehow in the matching code)

A lot of bugs are being exposed that should be fixed:
```
>           return _remove_gcd(x_0, y_0, z_0)
E           UnboundLocalError: local variable 'x_0' referenced before assignment
```

There are a few things in simplify etc that would need updating like collect/expand:
```
>       assert collect((1 + (x + y) + (x + y)**2).expand(),
                       [x, y]) == 1 + y + x*(1 + 2*y) + x**2 + y**2
E       assert x**2 + x + y**2 + y + 3 == ((((1 + y) + (x * (1 + (2 * y)))) + (x ** 2)) + (y ** 2))
```

The main substantive problem seems to be the lack of additive cancellation though:
```
>       assert reduce_inequalities(x + y < 1, symbols=[x]) == (x < 1 - y)
E       assert x + y - (y - 1) - 1 < -(y - 1) == (x < (1 - y))
```

I got to 70% complete with 549 failures out of 6000 before it hung in solvers somewhere. Rerunning only the failed tests a few pass on the second run which suggests some non-deterministic fails in there.

This doesn't look like an easy thing to fix but I really think it needs fixing. We need multiplication to be associative so `(a*b)*c = a*(b*c)`.

smichr committed:
4596: remove autoexpansion of 2-arg Mul

smichr committed:
fix core

smichr committed:
add Mul.__neg__

oscarbenjamin commented:
I think this is important but I haven't yet found time to look into this in detail.

I think we need to try and agree on some basic invariants to aim for like
```
a*b = b*a
(a*b)*c = a*(b*c)
-e = (-1)*e
a - e = a + (-1)*e = a + (-e)
```
and so on.

I thought about creating a database of expressions to test this. Perhaps we could patch Expr to save sreprs of all expressions to a file while running some part of the test suite and then run tests for a range of invariants across the examples to look for failures.

smichr committed:
almost WYSIWIG sympify

smichr committed:
Revert "almost WYSIWIG sympify"

This reverts commit ef1f3a0bea4e5a14ed6548c707875b7e6a6af648.

smichr commented:
Thanks @oscarbenjamin for the feedback in this new rendition.

```
a*b = b*a
(a*b)*c = a*(b*c)
```
This will be true because a Mul's args are sorted and there is no autodistribution. However, while `-1*a == a*-1` the crack that allows things to mostly work is that `-a != a*-1` when `a` is an Add (see below).

```
-e = (-1)*e
```
This is the point of contention (figuratively). Because Python supports the difference between negation and multiplication this allows us to keep most things working as one would expect if we let `-(a + b) == -a - b`.  It is easy enough *not* to work around this in `flatten` but much code would need to be rewritten to use a function or method to distinguish between negating terms vs multiplying by negative 1; I formally did this with a `neg` method. This comes up, for example, if you have something that has a negative coefficient and you want to make the coefficient positive and pass the negative on to the args as in `-2*(x + y) -> 2*(-x - y)` as when handling an exponent. It is not helpful to be left with `-(x + y)` because then you are back almost where you started: with a negative coefficient.
```
a - e = a + (-1)*e = a + (-e)
```
This behavior hinges on the above. While investigating this I see that unless we make `-e == -1*e` then we will still have an undesirable and uncontrollable Python parsing situation: we can define `__rsub__` to handle `x - (y + z)` and make that `x + -1*(y + z)` but if `x` is a Python primitive as in `1 - (x + y)` then Python negates the `x + y` and adds it to the 1 which SymPy converts to `S.One`...but the damage is already done and the negative sign is distributed.  

So it looks like the only way to live with those parsing rules is to make `-e == -1*e`...and to force all code doing `x = -x` to consider whether the negative should be distributed or not.

oscarbenjamin commented:
> So it looks like the only way to live with those parsing rules is to make `-e == -1*e`...and to force all code doing `x = -x` to consider whether the negative should be distributed or not.

I agree. I understand why Python distinguishes between 1-x and 1+(-x) but that distinction should not be relevant to SymPy users.

I don't think I've considered this as deeply as you have but it seems to me that there are lots of cases where we would like the minus signs to cancel like `(-x-y)*(-z-t)`. We need to make sure that it is done consistently and is associative e.g. `((-1-x)*(-1-y))*(-1-z) = (-1-x)*((-1-y)*(-1-z))`. There is a way to do that for example if both become `-((1+x)*(1+y)*(1+z))` in the end in this example. (Maybe I'm looking at a level just above what this PR is about here though...)

For this PR (or this work rather) I think we need a principled approach to which invariants are important and I think that some other input to that e.g. from @asmeurer @jksuom would be hugely beneficial because it would be a lot of work to see this to completion.

smichr commented:
> (-x-y)*(-z-t)

There is nothing in this PR to address this case. The goal of this PR is to stop a leading number from being distributed into a following Add. For the example you give, simplification already exists:

```python
>>> (-x-y)*(-x-z)
(-x - y)*(-x - z)
>>> signsimp(_)
(x + y)*(x + z)
```

smichr committed:
fix core

000-x is used to suppress the trap in __neg__ which raises
    an error when an Add is negated since formerly this would
    not have kept the sign factored out and now it does. So
    at each of these places a decision will have to be made
    as to whether negation or distribution of sign in the
    case of an Add should be made.

x.neg is used to distribute the sign when x is an Add else
    it just creates a leading negative, perha

    Not all instances of negation will have to use neg to
    distribute the
    sign, e.g. `if isinstance(-x, cls):` where `cls` is not Add
    will not need to use `neg` and can remain a negation.
    It could use negation, but there is no need to take the time to
    distribute the sign, sort the args and rebuild an Add
    which will fail the test anyway.

    The goal right now is to use the trap in __neg__ to bring to
    fore all places where negation is used so a decision can be
    made as to whether it is necessary or not.

oscarbenjamin commented:
> There is nothing in this PR to address this case.

Sorry I realise that. That's just something that I was thinking about since the change here affects when minus signs are distributed.

With this PR what would be the recommended way to simplify something like
```julia
In [18]: e                                                                                                                        
Out[18]: x + y - 2⋅(x + y)

In [19]: expand(e)                                                                                                                
Out[19]: -x - y

In [20]: cancel(e)                                                                                                                
Out[20]: -(x + y)
```

oscarbenjamin commented:
I see that's what x2 is for! Maybe that needs a better name...

smichr commented:
> Maybe that needs a better name...

I just pushed the most recent changes. I am using a trap in `__neg__` to alert to places executing `-foo` if `foo` can ever be an Add; there I am writing `000-foo`which then treats the negation as a subtraction. Eventually the `000` will be removed if it is determined that `foo.neg` is not needed.

So added methods are 
* `neg` to distribute a negative on terms: `x.neg` would be `-x` while `(x+y).neg` would become `-x - y` with args sorted canonically.
* `x2` to expand all Number*Add terms in an expression. It is not optimized, it just gets the job done.
* `__neg__` routines were added and optimized

All names can be changed. This is all just scaffolding to get from A to Z.

If you wouldn't mind taking a look at the `Add.flatten` docstring to see if the behavior makes sense. The motivation for allowing some cancellation to occur from within a negative term is that it is not uncommon to subtract away part of an expression. Without this behavior, we would always have to use expand in some form (perhaps `x2`) to get the terms to cancel.

jksuom commented on sympy/core/expr.py:
```diff
@@ -160,7 +160,16 @@ def __pos__(self):
+        # Mul has its own __neg__ routine so here, we just
+        # create a 2-args Mul with the -1 in the canonical
+        # slot 0
+        assert not self.is_Add
```
Is this expected:
```
>>> x = symbols('x')
>>> -(x + 1)*(x - 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/core/expr.py", line 166, in __neg__
    assert not self.is_Add
AssertionError
```

smichr replied:
During the detection/correction phase of the work. There is a trap to detect negated Adds right now. You can get around this (for now) by writing `0-(x+1)`.

smichr committed:
global_distribute controls negation behavior

smichr committed:
with distribute(True): -(x+1) -> -x-1

smichr committed:
fix algebras

smichr committed:
fix assumptions

smichr committed:
Add.neg flattens args

smichr committed:
fix algebras/assumptions

smichr committed:
fix some wester

smichr committed:
make sure neg handles unevaluated Add

smichr committed:
update neg docstring

asmeurer commented:
> I thought about creating a database of expressions to test this. Perhaps we could patch Expr to save sreprs of all expressions to a file while running some part of the test suite and then run tests for a range of invariants across the examples to look for failures.

hypothesis is really good for this sort of thing https://github.com/sympy/sympy/pull/17190. Although with anything we might test with hypothesis I think we should also take a known set of examples and add them to the test suite manually.

asmeurer commented:
> I don't think I've considered this as deeply as you have but it seems to me that there are lots of cases where we would like the minus signs to cancel like (-x-y)*(-z-t). We need to make sure that it is done consistently and is associative e.g. ((-1-x)*(-1-y))*(-1-z) = (-1-x)*((-1-y)*(-1-z)). There is a way to do that for example if both become -((1+x)*(1+y)*(1+z)) in the end in this example. (Maybe I'm looking at a level just above what this PR is about here though...)

A general issue with automatic distribution is that it isn't commutative. `2*x*(x + y)` becomes a different expression from `2*(x + y)*x`. I personally don't think minus signs in Adds should cancel automatically. That's the sort of thing that can easily be done in a simplification function (there are several functions that only extract common factors from Add terms).

asmeurer commented:
I know this is a draft but for draft PRs it's better to leave the release notes blank rather than "NO ENTRY" so that you don't forget to add them before merging.

smichr commented:
I am going to abandon this. Considering the amount of changes needed to just make a few modules pass (and spending a decent amount of time trying to get meijer/hyper to pass) I don't see the payoff. We have the ability for users to enter expressions literally using the `distribute` context:
```python
>>> from sympy.core.evaluate import distribute
>>> with distribute(False):
...  2*(3+y) - (x - y+z)
...
2*(y + 3) - (x - y + z)
```
Given this, what is the codebase advantage for making the change?

asmeurer commented:
Another cautionary tale that automatic evaluation can be nearly impossible to remove once it is added.

smichr commented:
Are you in favor of closing this issue as WontFix?

oscarbenjamin commented:
This seems worthwhile but also a *lot* of work so... I'm still in favour of someone (other than me) doing it :)

More seriously I wonder how much of the work to do this is actually fixing bugs that should be fixed anyway (not that that would make it any easier - probably the opposite in fact)

smichr commented:
> fixing bugs

My experience so far is that it isn't bugs, it's expectations. For example, if we started using an unordered list instead of a sorted list for args, everywhere where args[0] was assumed to be the location of a Number (if any) would no longer be true.

asmeurer commented:
You can close the PR as WontFinish. I would prefer to keep the issue open because I would like to see it someday, if someone can manage to do it. So maybe WontFixToday. Although you failing doesn't give me much hope, because I feel like you are one of the people who is willing to go the farthest in fixing things like this.

asmeurer commented:
> For example, if we started using an unordered list instead of a sorted list for args, everywhere where args[0] was assumed to be the location of a Number (if any) would no longer be true.

Incidentally, arg sorting is also something I'd like to see removed some day (we can keep Mul.args[0]), because it wastes a lot of time to do sorting.
